### PR TITLE
add counter reset support

### LINKED ISSUE #124: Support generating counter values
As of now, the `corpus-generator-tool` produces different `gauge` values with the use of `ranges`, `enum` and `fuzziness` as per the config [documentation](https://github.com/elastic/elastic-integration-corpus-generator-tool/blob/main/docs/fields-configuration.md).

However there can be a config defining a particular field as `counter` so it always produces incremental values. There are many such packages which are TSDB migrated which are using these `counter` [fields](https://github.com/elastic/integrations/blob/main/packages/apache_spark/data_stream/driver/fields/fields.yml#L24). It would be easier to implement the `rally` benchmarking templates as well.

- [ ] Counter support with no reset ([PR](https://github.com/elastic/elastic-integration-corpus-generator-tool/pull/129))
- [ ] Support for counter reset

Counters can be reset:
1. having the possibility to enable/disable just totally random occurrences of the reset.
2. having the possibility to define the probability that the counter is reset every time a new value is generated
3. having the possibility to specify the number of events after which the counter must be reset

Proposal for a `counter_reset` configuration.
```
1. 
counter_reset: 
  strategy: random
2.
counter_reset: 
  strategy: probability
  probability: 0-100 # %
3.
counter_reset: 
  strategy: events_amount
  events_amount: N
```

cc:- @aspacca

### PR DESCRIPTION
**Pull request by ghost**
### Overview

This PR introduces the counter reset functionality. It can be configured with 3 strategies: `random`, `probabilistic` and after a certain number (`after_n`).

Strategies:
- `random`
  - 50% chance to reset the counter
- `probabilistic`
  - Resets the counter based on a given probability which must be provided as a percentage (0-100).
- `after_n`
  - Resets the counter after a specified number of iterations (N). A new boolean was added to the state to check if the counter was reset. 

**Example:**

Random reset.

```yaml
  - name: test_counter
    counter: true
    counter_reset:
      strategy: "random"
```

Reset counter after 5 iterations.

```yaml
  - name: test_counter
    counter: true
    counter_reset:
      strategy: "after_n"
      reset_after_n: 5
```

20% chance to reset the counter.

```yaml
  - name: test_counter
    counter: true
    counter_reset:
      strategy: "probabilistic"
      probability: 20
```

### Test with config

<details>
<summary><strong>configs.yml</strong></summary>
<p>

```yaml
fields:
  - name: cloud.region
    enum: ["us-east-1", "us-east-2", "us-west-1", "us-west-2", "ap-south-1", "ap-northeast-3", "ap-northeast-2", "ap-southeast-1", "ap-southeast-2", "ap-northeast-1", "ca-central-1", "eu-central-1", "eu-west-1", "eu-west-2", "eu-west-3", "eu-north-1", "sa-east-1", "af-south-1", "ap-east-1", "ap-south-2", "ap-southeast-3", "eu-south-2", "eu-central-2", "me-south-1", "me-central-1"]
    cardinality: 25
  - name: cloud.account.id
    value: "123456789"
  - name: cloud.account.name
    value: sample-account
  - name: aws.billing.currency
    value: "USD"
  - name: aws.billing.ServiceName
    # NOTE: When empty the data refers to estimated charged for the entire account. We cannot reproduce the content (as it's a sum of previous data) but we want to provide the case.
    enum: ["", "AWSCloudTrail", "AWSCodeArtifact", "AWSConfig", "AWSCostExplorer", "AWSDataTransfer", "AWSELB", "AWSLambda", "AWSMarketplace", "AWSQueueService", "AWSSecretsManager", "AWSServiceCatalog", "AWSSystemsManager", "AWSXRay", "AmazonApiGateway", "AmazonCloudWatch", "AmazonCognito", "AmazonDynamoDB", "AmazonEC2", "AmazonECR", "AmazonEKS", "AmazonKinesis", "AmazonKinesisFirehose", "AmazonRDS", "AmazonRedshift", "AmazonRoute53", "AmazonS3", "AmazonSNS", "AmazonVPC", "awskms"]
  - name: agent.id
    value: "12f376ef-5186-4e8b-a175-70f1140a8f30"
  - name: agent.ephemeral_id
    value: "5fd278ce-2a12-4a09-a125-0c5b39aa69e3"
  - name: agent.name
    value: "host.local"
  - name: metricset.period
    value: 86400
  - name: aws.billing.group_definition.key
    # NOTE: repeated values are needed to produce 10% cases with "" value
    enum: ["", "AZ", "INSTANCE_TYPE", "SERVICE", "LINKED_ACCOUNT", "AZ", "INSTANCE_TYPE", "SERVICE", "LINKED_ACCOUNT"]

  - name: event.duration
    range:
      min: 1
      max: 1000
  - name: aws.billing.EstimatedCharges
    cardinality: 25
    fuzziness: 0.2
  - name: aws.billing.AmortizedCost.amount
    cardinality: 25
    fuzziness: 0.2
  - name: aws.billing.BlendedCost.amount
    cardinality: 25
    fuzziness: 0.2
  - name: aws.billing.NormalizedUsageAmount.amount
    cardinality: 25
    fuzziness: 0.2
  - name: aws.billing.UnblendedCost.amount
    cardinality: 25
    fuzziness: 0.2
  - name: aws.billing.UsageQuantity.amount
    cardinality: 25
    fuzziness: 0.2
  - name: aws.billing.group_definition.type
    value: "DIMENSION"
  - name: aws.billing.group_by.INSTANCE_TYPE
    enum: ["NoInstanceType", "a1.large", "c5.2xlarge", "c5.xlarge", "c6i.2xlarge", "db.r6g.2xlarge", "db.t2.micro", "dc2.large", "m5.large", "t1.micro", "t2.medium", "t2.micro", "t2.small", "t2.xlarge", "t3.2xlarge", "t3.medium", "t3.xlarge","t3.xlarge"]
  - name: aws.billing.group_by.SERVICE
    enum: ["Amazon Simple Storage Service", "Amazon Elastic Compute Cloud - Compute", "EC2 - Other", "Amazon Kinesis", "Amazon Relational Database Service", "Amazon Elastic Load Balancing", "AmazonCloudWatch", "AWS CloudTrail", "AWS Config", "AWS Key Management Service", "AWS Lambda", "AWS Secrets Manager", "AWS Service Catalog", "Amazon API Gateway", "Amazon DynamoDB", "Amazon EC2 Container Registry (ECR)", "Amazon Elastic Container Service for Kubernetes", "Amazon Kinesis Firehose", "Amazon Redshift", "Amazon Simple Notification Service", "Amazon Simple Queue Service", "Amazon Virtual Private Cloud"]
  - name: test_counter
    counter: true
    counter_reset:
      strategy: "after_n"
      reset_after_n: 5
```

</p>
</details> 

<details>
<summary><strong>fields.yml</strong></summary>
<p>

```yaml
- name: timestamp
  type: date
- name: cloud.region
  type: keyword
- name: cloud.account.id
  type: keyword
- name: cloud.account.name
  type: keyword
- name: event.duration
  type: long
- name: metricset.period
  type: long
- name: aws.billing.currency
  type: keyword
- name: aws.billing.EstimatedCharges
  type: float
  # positive
- name: aws.billing.ServiceName
  type: keyword
- name: aws.billing.AmortizedCost.amount
  type: float
  # positive
- name: aws.billing.BlendedCost.amount
  type: float
  # positive
- name: aws.billing.NormalizedUsageAmount.amount
  type: integer
  # positive
- name: aws.billing.UnblendedCost.amount
  type: float
  # positive
- name: aws.billing.UsageQuantity.amount
  type: integer
  # positive
- name: agent.id
  type: keyword
- name: agent.name
  type: keyword
- name: agent.ephemeral_id
  type: keyword
  example: 12f376ef-5186-4e8b-a175-70f1140a8f30
- name: aws.billing.group_definition.key
  type: keyword
- name: aws.billing.start_date
  type: date
- name: aws.billing.group_definition.type
  type: keyword
- name: aws.billing.group_by.INSTANCE_TYPE
  type: keyword
- name: aws.billing.group_by.SERVICE
  type: keyword
- name: test_counter
  type: long
```

</p>
</details> 

<details>
<summary><strong>gotext.tpl</strong></summary>
<p>

```
{{- $currency := generate "aws.billing.currency" }}
{{- $groupBy := generate "aws.billing.group_definition.key" }}
{{- $period := generate "metricset.period" }}
{{- $cloudId := generate "cloud.account.id" }}
{{- $cloudRegion := generate "cloud.region" }}
{{- $timestamp := generate "timestamp" }}
{
    "@timestamp": "{{$timestamp.Format "2006-01-02T15:04:05.999999Z07:00"}}",
    "cloud": {
        "provider": "aws",
        "region": "{{$cloudRegion}}",
        "account": {
            "id": "{{$cloudId}}",
            "name": "{{generate "cloud.account.name"}}"
        }
    },
    "event": {
        "dataset": "aws.billing",
        "module": "aws",
        "duration": {{generate "event.duration"}}
    },
    "metricset": {
        "name": "billing",
        "period": {{$period}}
    },
    "ecs": {
        "version": "8.2.0"
    },
    "aws": {
        "billing": {
{{- if eq $groupBy "" }}
            "Currency": "{{$currency}}",
            "EstimatedCharges": {{generate "aws.billing.EstimatedCharges"}},
            "ServiceName": "{{generate "aws.billing.ServiceName"}}"
{{- else }}
{{- $sd := generate "aws.billing.start_date" }}
            "start_date": "{{ $sd.Format "2006-01-02T15:04:05.999999Z07:00" }}",
            "end_date": "{{ $sd | date_modify (print "+" $period "s") | date "2006-01-02T15:04:05.999999Z07:00" }}",
            "AmortizedCost": {
                "amount": {{printf "%.2f" (generate "aws.billing.AmortizedCost.amount")}},
                "unit": "{{$currency}}"
            },
            "BlendedCost": {
                "amount": {{printf "%.2f" (generate "aws.billing.BlendedCost.amount")}},
                "unit": "{{$currency}}"
            },
            "NormalizedUsageAmount": {
                "amount": {{generate "aws.billing.NormalizedUsageAmount.amount"}},
                "unit": "N/A"
            },
            "UnblendedCost": {
                "amount": {{printf "%.2f" (generate "aws.billing.UnblendedCost.amount")}},
                "unit": "{{$currency}}"
            },
            "UsageQuantity": {
                "amount": {{generate "aws.billing.UsageQuantity.amount"}},
                "unit": "N/A"
            },
            "group_definition": {
              "key": "{{$groupBy}}",
              "type": "{{generate "aws.billing.group_definition.type"}}"
            },
            "test_counter": {{generate "test_counter"}},
            "group_by": {
{{- if eq $groupBy "AZ"}}
              "AZ": "{{awsAZFromRegion $cloudRegion}}"
{{- else if eq $groupBy "INSTANCE_TYPE"}}
              "INSTANCE_TYPE": "{{generate "aws.billing.group_by.INSTANCE_TYPE"}}"
{{- else if eq $groupBy "SERVICE"}}
              "SERVICE": "{{generate "aws.billing.group_by.SERVICE"}}"
{{- else if eq $groupBy "LINKED_ACCOUNT"}}
              "LINKED_ACCOUNT": "{{$cloudId}}"
{{- end}}
            }
{{- end}}
        }
    },
    "service": {
        "type": "aws"
    },
    "agent": {
        "id": "{{generate "agent.id"}}",
        "name": "{{generate "agent.name"}}",
        "type": "metricbeat",
        "version": "8.0.0",
        "ephemeral_id": "{{generate "agent.ephemeral_id"}}"
    }
}
```

</p>
</details> 

```sh
go run main.go generate-with-template ./gotext.tpl ./fields.yml --config-file ./configs.yml --tot-events 10
```

Relates: #141 
Closes: #124

### CODE REVIEW

**Review by Oscurz — APPROVED** [review]

**Comment by shmsr** [comment]
Can you also add documentation or examples so that people can refer to the same? These are kind of features that needs to be thoroughly documented so that behavior is understandable.

Also, few comments:

- Add validation for:
  -  `probabilistic`: Validate during config parsing i.e., the `probability` value is in [0, 100]
  - `after_n`: Validate during config parsing i.e., `reset_after_n` value is more than 0
- You have added this logic for `bindLongWithReturn`, why not also add it for `bindDoubleWithReturn`. Very nice feature to have.
- The current implementation resets the counter to 0. What do you think of having an optional field to configure reset value? Could be useful.

File: pkg/genlib/generator_interface.go [hunk 1]
```diff
@@ -1008,6 +1019,27 @@ func bindLongWithReturn(fieldCfg ConfigField, field Field, fieldMap map[string]a
...
+					// Probability% chance to reset
+					if customRand.Intn(100) < int(*fieldCfg.CounterReset.Probability) {
+						dummyInt = 0
+					}
+				case config.CounterResetStrategyAfterN:
+					// Reset after N
+					if !state.counterReset && state.counter >= *fieldCfg.CounterReset.ResetAfterN {
+						dummyInt = 0
+						state.counterReset = true
+					}
```
Comment (shmsr) [inline pkg/genlib/generator_interface.go:1039]: After resetting once, it is always set to reset. A bug?

```go
} else if state.counter < *fieldCfg.CounterReset.ResetAfterN {
	state.counterReset = false
}
```

Add this logic?

**Comment by shmsr** [comment]
Also, if a critical change is done here, point this PR to the main issue itself. So that others can also discuss if we are implemented as expected?

File: pkg/genlib/generator_with_text_template_test.go [hunk 2]
```diff
@@ -755,6 +755,64 @@ func Test_FieldIPWithTextTemplate(t *testing.T) {
 	}
 }
 
+func Test_FieldLongCounterResetAfterN5WithTextTemplate(t *testing.T) {
```
Comment (shmsr) [inline pkg/genlib/generator_with_text_template_test.go:758]: Add cases where the counter is incrementing properly between multiple resets to consider the case in my previous comment.

File: pkg/genlib/generator_interface.go (hunk 1 above)
Comment (shmsr) [reply pkg/genlib/generator_interface.go:1039]: Okay, then it is fine.

Comment (gpop63) [reply pkg/genlib/generator_interface.go:1039]: Do you think it should be every N iterations or just once like it is now?

Comment (shmsr) [reply pkg/genlib/generator_interface.go:1039]: IMO, every. If it is happening once, then why not every `n`.

Let's ask @aliabbas-elastic as well if he thinks that would be useful to have.

**Review by shmsr — APPROVED** [review]

File: docs/fields-configuration.md [hunk 3]
```diff
@@ -15,6 +15,15 @@ For each config entry the following fields are available:
...
+- `counter_reset` *optional (only applicable when `counter: true`)*: configures how and when the counter should reset. It has the following sub-fields:
+  - `strategy` *mandatory*: defines the reset strategy. Possible values are:
+      - `"random"`: resets the counter at random intervals.
+      - `"probabilistic"`: resets the counter based on a probability.
+      - `"after_n"`: resets the counter after a specific number each iteration.
+  - `probability` *required when strategy is "probabilistic"*: an integer between 1 and 100 representing the percentage chance of reset for each generated value.
+  - `reset_after_n` *required when strategy is "after_n"*: an integer specifying the number of values to generate before resetting the counter.
+
+Note: The `counter_reset` configuration is only applicable when `counter` is set to `true`.
```
Comment (shmsr) [inline docs/fields-configuration.md:26]: this is already mentioned in line 18 too; do you this is needed?

File: docs/fields-configuration.md [hunk 4]
```diff
@@ -15,6 +15,15 @@ For each config entry the following fields are available:
...
+- `counter_reset` *optional (only applicable when `counter: true`)*: configures how and when the counter should reset. It has the following sub-fields:
+  - `strategy` *mandatory*: defines the reset strategy. Possible values are:
+      - `"random"`: resets the counter at random intervals.
+      - `"probabilistic"`: resets the counter based on a probability.
+      - `"after_n"`: resets the counter after a specific number each iteration.
+  - `probability` *required when strategy is "probabilistic"*: an integer between 1 and 100 representing the percentage chance of reset for each generated value.
+  - `reset_after_n` *required when strategy is "after_n"*: an integer specifying the number of values to generate before resetting the counter.
```
Comment (shmsr) [inline docs/fields-configuration.md:24]:
```suggestion
- `counter_reset` (optional, only applicable when `counter: true`): Configures how and when the counter should reset. It has the following sub-fields:
  - `strategy` (required): Defines the reset strategy. Possible values are:
    - `"random"`: Resets the counter at random intervals.
    - `"probabilistic"`: Resets the counter based on a probability.
    - `"after_n"`: Resets the counter after a specific number of iterations.
  - `probability` (required when `strategy: "probabilistic"`): An integer between 1 and 100 representing the percentage chance of reset for each generated value.
  - `reset_after_n` (required when `strategy: "after_n"`): An integer specifying the number of values to generate before resetting the counter.
```

**Pull request merged**